build/i2c/src/lib.rs

```diff
@@ -1796,3 +1821,46 @@ where
     }
     Ok(())
 }
+
+impl Refdes {
+    fn to_component_id(&self) -> String {
+        self.join_with_case(str::make_ascii_uppercase, "/")
+    }
+
+    fn to_upper_ident(&self) -> String {
+        self.join_with_case(str::make_ascii_uppercase, "_")
+    }
+
+    fn to_lower_ident(&self) -> String {
+        self.join_with_case(str::make_ascii_lowercase, "_")
+    }
+
+    fn join_with_case(
+        &self,
+        change_case: impl Fn(&mut str),
+        sep: &str,
+    ) -> String {
+        match self {
+            Self::Component(s) => {
+                let mut s = s.clone();
+                change_case(&mut s);
+                s
+            }
+            Self::Path(parts) => {
+                let len = parts.iter().map(String::len).sum::<usize>()
+                    + (parts.len() - 1);
```

Comment:
```suggestion
                    + (parts.len() - 1) * sep.len();
```
(heaven forbid we have to reallocate if someone specifies a multi-character separator)